### PR TITLE
Add QuaternionTest.mpl; Fix [!], style and typos in Quaternion.mpl

### PR DESCRIPTION
**Pull request by 3DpoNCh1k**
Add tests for quaternions and fix some typos in Quaternion.mpl

### CODE REVIEW

File: tests/QuaternionTest.mpl [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+[
+  q0: (1.0 2.0 neg 3.0 4.0    ) quaternion unit;
+  q1: (5.0 6.0     7.0 8.0 neg) quaternion unit;
+  tests: (
+    {f: 0.0; q: q0 new                                                                                      ;}
+    {f: 0.3; q: (0.000501537327541 neg 0.48176120346404 neg 0.23987752707694     0.8428313373983) quaternion;}
+    {f: 0.6; q: (0.183499748964156 neg 0.52391886713953 neg 0.10504006435854 neg 0.8251081430121) quaternion;}
+    {f: 1.0; q: q1 new                                                                                      ;}
+  );
+  2 [
```
Comment (Parean) [inline tests/QuaternionTest.mpl:116]: Remove copypaste

File: tests/QuaternionTest.mpl [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    tests [
+      t:;
+      q0 q1 t.f              slerp t.q checkEqual
+      q1 q0 1 t.f cast t.f - slerp t.q checkEqual
+    ] each
+    q1.entries neg @q1.!entries
+  ] times
+] call
+
+# matrix
```
Comment (Parean) [inline tests/QuaternionTest.mpl:126]: These comments aren't very informative. See, for example, [this ](https://github.com/Matway/mpl-sl/blob/master/tests/ContextTest.mpl#L24) as a good reference

File: tests/QuaternionTest.mpl [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+];
+
+# basic ops
+[
+  q0: (1.0 2.0 neg 3.0 4.0    ) quaternion;
+  q1: (5.0 6.0     7.0 8.0 neg) quaternion;
+
+  q: (0.182574185835055 0.365148371670111 neg 0.5477225575051 0.7302967433402) quaternion;
+  tests: (
+    {op: [q0 q1 +  ]; expected: ( 6.0  4.0     10.0      4.0 neg) quaternion;}
```
Comment (Parean) [inline tests/QuaternionTest.mpl:67]: What about "!"?

Comment (Parean) [reply tests/QuaternionTest.mpl:67]: Also, I would expect some tests with identityQuaternion multiplication

Comment (Parean) [reply tests/QuaternionTest.mpl:67]: Also, I don't see tests for quaternionCast

**Review by Parean — CHANGES_REQUESTED** [review]

File: tests/QuaternionTest.mpl [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+"algebra.unit"    use
+"algebra.trans"   use
+"algebra.vector?" use
+"algorithm.each"  use
+"control.abs"     use
+"control.ensure"  use
+"control.pfunc"   use
+"control.times"   use
+
+"Quaternion" use
```
[the remark this answers is not in the document]
Comment (3DpoNCh1k) [reply tests/QuaternionTest.mpl:22]: even if I need many functions defined in Quaternion file? I did as in algorithmTest

File: tests/QuaternionTest.mpl (hunk 2 above)
Comment (3DpoNCh1k) [reply tests/QuaternionTest.mpl:126]: I used them to name individual test functions, as in algorithmTest. I can add comments to some subparts of tests where some properties are checked, but I think they are clear without comments

File: tests/QuaternionTest.mpl (hunk 1 above)
Comment (3DpoNCh1k) [reply tests/QuaternionTest.mpl:116]: I can combine slerp and nlerp into one test function and extend tests structures like
```
tests: (
  {
   f:;
    q: {
     nlerp:;
     slerp:;
    }
  }
) 
```
Is it ok?

File: tests/QuaternionTest.mpl (hunk 2 above)
Comment (Parean) [reply tests/QuaternionTest.mpl:126]: I think that some simple properties like the commutativity of quaternion multiplication can be omitted. But more or less non-trivial properties are worth pointing out.

File: tests/QuaternionTest.mpl (hunk 1 above)
Comment (Parean) [reply tests/QuaternionTest.mpl:116]: I think it´s better to use helper function.

File: tests/QuaternionTest.mpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (Parean) [reply tests/QuaternionTest.mpl:22]: Yes, even in that case. Code in algorithmTest should be fixed.

**Review by Parean — CHANGES_REQUESTED** [review]

File: tests/QuaternionTest.mpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (Parean) [reply tests/QuaternionTest.mpl:22]: Yes, even in that case. Code in algorithmTest should be fixed.

**Review by Parean — APPROVED** [review]

**Pull request merged**